eng/common/scripts/modules/ChangeLog-Operations.psm1

```diff
@@ -0,0 +1,109 @@
+# Common Changelog Operations
+
+$RELEASE_TITLE_REGEX = "(?<releaseNoteTitle>^\#+.*(?<version>\b\d+\.\d+\.\d+([^0-9\s][^\s:]+)?)\s(?<releaseStatus>\(Unreleased\)|\(\d{4}-\d{2}-\d{2}\))?)"
+
+# given a CHANGELOG.md file, extract the relevant info we need to decorate a release
+function Extract-ReleaseNotes {
+    param (
+        [Parameter(Mandatory = $true)]
+        [String]$ChangeLogLocation,
+        [String]$VersionString
+    )
+    
+    $ErrorActionPreference = 'Stop'
+
+    $releaseNotes = @{}
+    if ($ChangeLogLocation.Length -eq 0)
+    {
+        return $releaseNotes
+    }
+    
+    try 
+    {
+        $contents = Get-Content $ChangeLogLocation
+    
+        # walk the document, finding where the version specifiers are and creating lists
+        $version = ""
+        foreach($line in $contents){
+            if ($line -match $RELEASE_TITLE_REGEX)
+            {
+               if ($version -ne "")
+               {
+                  $releaseNotes[$version].ReleaseContent = $releaseContentBuilder -join [Environment]::NewLine
+                  $releaseNotes[$version].ReleaseNotesEntry = $releaseContentBuilder[1..($releaseContentBuilder.Length - 1)] -join [Environment]::NewLine
+               }
+               $version = $matches["version"]
+               $releaseContentBuilder = @()
+               $releaseNotes[$version] = New-Object PSObject -Property @{
```

Comment:
Instead of indexing all the time I'd suggest we simply store the releaseNote object in a local variable instead of the $version. Then if that isn't null then we += the releaseContent and releaseNotesEntry.

Reply:
Am not sure what you mean here. The version is being used as the key for the HashTable that contains the PSPbjects.
That still need to be indexed every time $line matches $RELEASE_TITLE_REGEX.

Reply:
I was thinking about simplifying this to something like:

```
$contents = Get-Content $ChangeLogLocation

$releaseNotesEntry = $null

foreach($line in $contents) {
  if ($line -match $RELEASE_TITLE_REGEX)
  {
     $releaseNotesEntry = [pscustomobject]@{ 
        ReleaseVersion = $matches["version"]
        ReleaseStatus = $matches["releaseStatus"]
        ReleaseTitle = $line
        ReleaseContent = @() # Release content without the version title
     }
     $releaseNotes[$releaseNotesEntry.ReleaseVersion] = $releaseNotesEntry
  }
  else {
    if ($releaseNotesEntry) {
      $releaseNotesEntry.ReleaseContent += $line
     }
  }
}
```

If there are still places that that need things to be joined by a newline then they can easily do that when they consume the information.

Reply:
Thanks a lot